Add Python bindings for libnetplan

slyon opened the pull request:
## Description
Implementation of public Python bindings for `libnetplan.so`, using [CFFI](https://cffi.readthedocs.io/) in API mode, as described in spec FO0114. The diff is fairly big, but it contains lots of refactorings, effectively moving 700 LOC of our old/internal ctypes-bindings over into a public module and porting them to make use of CFFI. Going through the commits one by one should make the review manageable:

8184c2b5 parse-nm: improve some error passing
59ee33cf CI: Add python3-cffi build-dependency
032e3568 include: Extend public API functions
9a84892a src: make internal API functions/structures available
4bfe29be bindings: python-cffi build skeleton
fdad4e08 bindings: Initial implementation of CFFI Python bindings
40c4f47f cli:tools:test: Make use of new cffi python bindings
a21d9f43 cli: drop legacy/internal ctypes bindings, now unused
8fd5adf2d1541adb8bd7a30fa3fea72e78ed4bf9 examples: Add an example of how to use Netplan's CFFI bindings

The most interesting bits are porbably those commits, where the new module is introduced, to be installed into `usr/lib/python3/dist-packages/netplan/`:
* https://github.com/canonical/netplan/commit/4bfe29be6b00f6ecf7b6dec6dc6351122a2a83ec bindings: python-cffi build skeleton
* https://github.com/canonical/netplan/commit/fdad4e08956bb0e29edf4d5ade9338d4d2c91312 bindings: Initial implementation of CFFI Python bindings

We're using the file in `python-cffi/netplan/_build_cffi.py` to build a `_netplan_cffi.cpython-311-x86_64-gnu.so` (internal) binary module, that is linked to `libnetplan.so` itself. In the `ffibuilder.cdef(...)` we declare any functions/constants/structures that we want to make available from libnetplan.so to the Python bindings. The compiler will match those against Netplan's headers at build time. In `python-cffi/netplan/{parser,state,netdef}.py` we're then making use of those declarations to define our high-level, Python-native bindings.

We're still making some limited use of ctypes in our tests (`test_libnetplan.py`) in order to validate some raw library functionality, which are not necessarily exposed through the Python bindings.

## Checklist

- [x] Runs `make check` successfully.
- [x] Retains 100% code coverage (`make check-coverage`).
- [ ] New/changed keys in YAML format are documented.
- [x] \(Optional\) Adds example YAML for new feature.
- [x] \(Optional\) Closes an open bug ~~in Launchpad~~: FR-2780

daniloegea commented on python-cffi/netplan/state.py:
```diff
@@ -0,0 +1,128 @@
+    def netdefs(self) -> NetDefinitionIterator:
+        return dict((nd.id, nd) for nd in NetDefinitionIterator(self, None))
+
+    # TODO: veths
```
suggestion: `virtual-ethernets` and `dummy-devices` could be added now.

daniloegea commented on examples/cffi-bindings.py:
```diff
@@ -0,0 +1,58 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
```
suggestion: it would be nice to have a short comment explaining what this script does.

slyon commented on python-cffi/netplan/netdef.py:
```diff
@@ -0,0 +1,185 @@
+        if not next_value:
+            raise StopIteration
+        content = next_value
+        # XXX: Introduce getters for .address/.lifetime/.label, to avoid
```
Yes.. we should build proper API around it, eventually. And keep it internal until that's done.

slyon commented:
Thanks @daniloegea! I've rebased and addressed most of your remarks. I still need to improve my comments/description in  `examples/cffi-bindings.py` and `python-cffi/netplan/meson.build` but maybe you could already double-check the logical changes I did so far.

daniloegea commented on gcovr.cfg:
```diff
@@ -0,0 +1 @@
+exclude = _build-cov/python-cffi/netplan/_*
```
I just realized that the test coverage-c is failing for me. The reason is because my build directory is called `build`, and not `_build-cov`... we will probably need to invert the logic here to not depend on the build directory name.

I tested this and it worked:

```
filter = src/*
filter = tests/ctests/*
```

daniloegea reviewed (approved):
Looks good!

Only one thing I'd like to be changed before merging: the new `gcovr.cfg` file probably should use the `filter` configuration to include the source code instead of `exclude` with a hardcoded build directory. The test coverage-c will fail if you use a different build directory.

slyon commented on python-cffi/netplan/meson.build:
```diff
@@ -0,0 +1,55 @@
+# See https://github.com/grimme-lab/xtb-python/blob/main/meson.build
+cffi_pyext = python.extension_module(
+    '_netplan_cffi',
+    link_whole: static_library(
```
After further investigation, it looks like we don't need that quirk at all, but can just build the binary Python module (shared object) directly from the source (see https://github.com/mesonbuild/meson/issues/5797).

slyon commented:
Thanks, I've addressed all of your comments. CI failures about Azure kernels are unrelated:
`E: Package 'linux-modules-extra-5.15.0-1042-azure' has no installation candidate`

Let's merge this!
